### PR TITLE
Support for dynamic subscriptions

### LINKED ISSUE #814: Proposal: Support dynamic (on-demand) subscribe and unsubscribe of topics by user app
Currently dapr calls into user app on route /dapr/subscribe to get a list of topics the app is interested in. For some apps the list of topics which they want to subscribe might no be known at start up time and can be decided later based on the business logic in app.
So Dapr should support dynamic subscription of topics by user app. App should be able to call an endpoint to subscibe (and to unsubscribe) for a topic. 

Things to consider:
In current model, after crash dapr runtime can call into user app to get topic for subscription, but with dynamic subscribe, dapr runtime since there will be on url to call to get list of topics, there should  be another way to get subscribed topics from app after dapr rutnime crash. 

Note: Docs PR https://github.com/dapr/docs/issues/2404at the time of code checkin

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
# Description

As the last part of the journey that started with https://github.com/dapr/components-contrib/pull/1756 , added the ability to subscribe to and unsubscribe from topics on-demand.

This PR adds 3 new APIs:

- HTTP:
   - `GET  /v1.0/subscriptions` -> returns the list of topics (and routes, etc) the application is subscribed to
   - `POST /v1.0/subscriptions` -> subscribes to a new topic (on an existing pubsub component)
   - `DELETE /v1.0/subscriptions` -> unsubscribes from an existing topic subscription
- gRPC equivalents:
   - `ListActiveTopicSubscriptions`
   - `SubscribeTopic`
   - `UnsubscribeTopic`

Altogether, they allow developers to subscribe to new topics on-demand, and unsubscribe when not needed anymore.

The biggest concern here is that the list of active subscriptions is not persisted anywhere, so in case `daprd` crashed (and the app stayed active), that would be lost. This is how we are addressing that problem:

- Applications are recommended to continue implementing the callback to return the initial list of subscriptions (in the case of HTTP, that would be `/dapr/subscribe`), returning an empty list if appropriate
- Methods to subscribe to and unsubscribe from a topic return in the response the full list of active subscriptions. Thanks to the use of locks, no more than one concurrent call to subscribe/unsubscribe could happen at the same time, so it's guaranteed that the response will contain the exact list of topics the runtime is subscribed to at the time it's sent.
- Apps are recommended to maintain the list of subscriptions (ie. the response from the subscribe/unsubscribe methods) and serve that on subsequent invocations to `/dapr/subscribe`.
   - This is something that the developers need to do, but we could potentially add that to the Dapr SDKs
- If the runtime crashed, but not the app, when the runtime comes back online it will query the `/dapr/subscribe` endpoint (or equivalent gRPC) and the app should respond with the last list of subscriptions

## Issue reference

Fixes #814

## Docs

TODO @ItalyPaleAle 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Unit tests passing
* [X] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by artursouza** [comment]
@ItalyPaleAle Please, move this to DRAFT and create a proposal to spark the discussion. I think this is a big change and the PR can help create clarity. I just think the discussion should be in an issue.

Thanks for this!

**Comment by ItalyPaleAle** [comment]
I will write down the details in the existing issue: https://github.com/dapr/dapr/issues/814

**Comment by yaron2** [comment]
> I will write down the details in the existing issue: #814

Thanks @ItalyPaleAle. We encourage setting expectations before submitting a PR as laid out in our [contributions guide](https://github.com/dapr/dapr/blob/master/CONTRIBUTING.md#pull-requests).

**Comment by yaron2** [comment]
This PR resolves an issue for 1.10, and ideally we'd want to see this make it to 1.10. @ItalyPaleAle do you think this is doable or shall we move to 1.11?

**Comment by ItalyPaleAle** [comment]
@yaron2 still waiting on your feedback on the issue since you had concerns on how this was designed.

**Comment by artursouza** [comment]
Can we revive this and separate unsubscribe into a separate PR?

**Comment by timgabrhel** [comment]
> Can we revive this and separate unsubscribe into a separate PR?

@artursouza Is any form of dynamic subscriptions available in a release yet? This PR is a little hard to follow.

**Comment by artursouza** [comment]
No. It is not. We need an API for the sdks to initialize the pubsub subscription. It does not exist. I don't mind having a fresh new PR.

**Comment by ItalyPaleAle** [comment]
This PR would need to be redone pretty much from scratch at this point. It was made before bulk pubsub and a lot other changes.

**Comment by danielgerlag** [comment]
@ItalyPaleAle @yaron2 @artursouza any news on this functionality?
This is a show stopper for me, if this cannot be supported by dapr, I will have to strip dapr pubsub from my project and use an underlying broker directly.

**Comment by artursouza** [comment]
It is not in our priority list (yet). We can revisit it but also accept a contribution.

**Comment by omidb** [comment]
@artursouza and @ItalyPaleAle is there any timeline to get this in? We really need to dynamically change the routes and topics we un/subscribe

**Comment by artursouza** [comment]
At this point, we don't have a timeline. We can revisit this once we are done with 1.13. If there is a contributor interested in implementing this feature, I can work with that person throughout.
